drivers: ipm driver based on stm32 HSEM

cameled opened the pull request:
Some STM32 SOC, like stm32h745 and stm32h747 doesn't have IPCC.
Provide an STM32 HSEM based ipm driver for these SOC.

erwango commented:
@cameled Great thanks for this contribution! 
To ease the review, is there a way we can test this ?
Ideally providing a way to build and test this driver in tree (so that at least this could be built by CI) would be a great addition.
Maybe a very basic sample that could be flashed on both cores would suffice. Anyway, this won't be a blocker as dual core testing is a particular case.

cameled commented:
> @cameled Great thanks for this contribution!
> To ease the review, is there a way we can test this ?
> Ideally providing a way to build and test this driver in tree (so that at least this could be built by CI) would be a great addition.
> Maybe a very basic sample that could be flashed on both cores would suffice. Anyway, this won't be a blocker as dual core testing is a particular case.

erwango commented:
> @cameled cameled closed this 42 seconds ago

@cameled, I'm sorry you closed this PR. Hope this is not definitive.
Please note that my request about adding a test is not blocking due to the specific dual core condition.
This can be done as a later steps and I'm sure we'll find help in the community to add it.

cameled commented:
Sorry, I think it close the comment, but it close the whole pull request.

cameled commented:
I'm preparing an simple test example now.
I have a stm32h747i_disco board, so example will work on it.

ABOSTM commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -737,6 +737,15 @@
+			compatible = "st,stm32-hsem-mailbox";
+			reg = <0x58026400 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_AHB4 0x02000000>;
+			interrupts = <125 0>;
```
As interrupt 125 is valid for Cortex M7 only, could it be moved to _m7 specific file ?

cameled replied:
OK, I will try to move it to _m7, same like _m4.

ABOSTM reviewed (changes requested):
@cameled, great job thanks.
2 minor remarks otherwise, LGTM

cameled commented on dts/arm/st/h7/stm32h7.dtsi:
```diff
@@ -737,6 +737,15 @@
 			#io-channel-cells = <1>;
 		};
 
+		mailbox: mailbox@58026400 {
```
OK, I will move it to stm32h7_dualcore.dtsi

mbolivar-nordic commented on drivers/ipm/ipm_stm32_hsem.c:
```diff
@@ -0,0 +1,197 @@
+	},
+};
+
+DEVICE_DT_INST_DEFINE(0,
```
Is there a good reason why this does not respect `status = "disabled";` by using DT_INST_FOREACH_STATUS_OKAY?

cameled replied:
STM32 HSEM has its own LL_HSEM(low-level HSEM) API provided by the hal_stm32 module. LL_HSEM API worked very well for the hardware semaphore mechanism.
The STM32 HSEM IPM driver I add here, Only pick up two semaphore ID from stm32_hsem.h to simulate a virtual mailbox device. 
So there will have only one instance.

mbolivar-nordic replied:
Having a single instance does not excuse ignoring status 🙂. I think the default stance should be to respect it.

cameled replied:
Sorry, misunderstanding the question, will add the 'status' check.

cameled replied:
I have added the DT_INST_FOREACH_STATUS_OKAY.

erwango commented on boards/arm/stm32h747i_disco/Kconfig.defconfig:
```diff
@@ -12,6 +12,14 @@ config BOARD
 config STM32H7_DUAL_CORE
 	default y
 
+if IPM_STM32_HSEM
```
Please also update board documentation

cameled replied:
OK, I will update the board documentation.

erwango reviewed (commented):
Thanks again for the great work.

Ideally, we'd request to split in various commits:
- dtsi/bindings
- soc
- driver
- board
Though this is not blocking.

Thanks for sharing the sample. It would be nice to have it under `samples/boards/stm32/h7_dual_core` so people can easily test it and make sure the driver is compiled in CI.

EDIT: I've performed the review in comments instead of requesting changes, as I will be out for some time and I would not want to block merging.

erwango commented:
@cameled Out of curiosity, are you planning to use openamp on top of this driver ?

cameled commented:
> @cameled Out of curiosity, are you planning to use openamp on top of this driver ?
Yes, I'm preparing to use openamp.

cameled commented:
It's my first-time to contribute code for an open source project.
If there have some unsuitable things, please let me know. 
Thanks a lot!

cameled commented on drivers/ipm/ipm_stm32_hsem.c:
```diff
@@ -0,0 +1,210 @@
+#define HSEM_CPU2                   2
+
+#if CONFIG_IPM_STM32_HSEM_CPU == HSEM_CPU1
+#define LL_HSEM_EnableIT_CIER       LL_HSEM_EnableIT_C1IER
+#define LL_HSEM_DisableIT_CIER      LL_HSEM_DisableIT_C1IER
+#define LL_HSEM_ClearFlag_CICR      LL_HSEM_ClearFlag_C1ICR
+#define LL_HSEM_IsActiveFlag_CISR   LL_HSEM_IsActiveFlag_C1ISR
+#else /* HSEM_CPU2 */
+#define LL_HSEM_EnableIT_CIER       LL_HSEM_EnableIT_C2IER
+#define LL_HSEM_DisableIT_CIER      LL_HSEM_DisableIT_C2IER
+#define LL_HSEM_ClearFlag_CICR      LL_HSEM_ClearFlag_C2ICR
+#define LL_HSEM_IsActiveFlag_CISR   LL_HSEM_IsActiveFlag_C2ISR
```
Yes, you are right. to avoid confusion with STM32 HSEM low-level API, I will change the redefined function name.

erwango commented:
> It's my first-time to contribute code for an open source project.

And you're doing great!

> If there have some unsuitable things, please let me know.

Nothing is unsuitable, but please address the inline comments provided above.

cameled commented:
@erwango The change will be sent Monday or Tuesday on next week.

mbolivar-nordic reviewed (approved):
Thanks for the update and for addressing my change request. Welcome to the project!

cameled commented on drivers/ipm/Kconfig:
```diff
@@ -112,6 +112,20 @@ config IPM_INTEL_ADSP
+config IPM_STM32_HSEM
+	bool "STM32 HSEM controller"
+	depends on STM32H7_DUAL_CORE
+	help
+		Driver for stm32 HSEM mailbox
```
Thanks.

cameled commented on drivers/ipm/Kconfig:
```diff
@@ -112,6 +112,20 @@ config IPM_INTEL_ADSP
+	default 1
+	range 1 2
+	depends on IPM_STM32_HSEM
+	help
+		use to define the CPU ID used by HSEM
```
Thanks.

Commit:
soc: arm: st_stm32: update HSEM ID for stm32 hsem ipm

New hardware semaphore ID for inter-processor mailbox

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

Commit:
drivers: ipm: add driver based on stm32 hsem

Some STM32 SOC, like stm32h745 and stm32h747 doesn't have IPCC.
Provide a STM32 HSEM based ipm driver for these SOC.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

Commit:
dts: stm32h7: define mailbox controller based on stm32 hsem

The ipm_stm32_hsem driver is a virtual mailbox based on STM32 HSEM.
Since we already have LL_HSEM(low-level HSEM) API on the hal_stm32
module, looks there is no need to implement a new HSEM driver.
But there are still have some limitations, the HSEMx interrupt is
not shareable. If another HSEM user wants to use the HSEMx interrupt,
the ipm_stm32_hsem mailbox needs to be disabled.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

Commit:
boards: arm: stm32h747i_disco: add ipm support

Add IPM support for stm32h747i_disco.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

Commit:
CODEOWNERS: add owners for stm32 hsem ipm driver

Add condeowner for STM32 HSEM IPM driver files.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

Commit:
samples: boards: stm32: add stm32 hsem ipm driver sample

Add a sample for stm32 hsem ipm driver.
Blinky led triggered by mailbox new message.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

cameled commented:
Sample have some problem. create a issue to track it. #37395

ABOSTM reviewed (approved):
LGTM
